File: Data Structures & Algorithms/depth-of-binary-tree/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0
        st = [(root, 1)]
        maxD = 0

        while st:
            node, currD = st.pop()
            maxD = max(currD, maxD)

            if(node.right):
                st.append((node.right, currD + 1))
            if(node.left):
                st.append((node.left, currD + 1))
                
        return maxD

File: Data Structures & Algorithms/valid-binary-search-tree/submission-2.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    bool helperValidBST(TreeNode* root, long long minVal, long long maxVal){
      if(!root)
        return true;
      
      if(root->val >= maxVal || root->val <= minVal){ return false; }
      return helperValidBST(root->left, minVal, root->val) && helperValidBST(root->right, root->val, maxVal);
    }
    bool isValidBST(TreeNode* root) {
        return helperValidBST(root, INT_MIN, INT_MAX);
    }
};
